File: python/ecuaciones_no_lineales/secante.py
import math
import matplotlib.pyplot
from sympy import *


def secante(f, xo, xi, tol, iterMax):
    x = Symbol('x')
    f1 = sympify(f)
    error = tol + 1
    k = 0
    it = []
    e = []
    while tol < error and k < iterMax:
        nume = (xi - xo)
        deno = f1.evalf(subs={x: xi})-f1.evalf(subs={x: xo})

        if deno == 0:
            it = []
            e = []
            print("La funcion se indefine.")
        else:
            xn = xi - ((nume) / (deno)) * ((f1.evalf(subs={x: xi})))
            error = abs(xn)
            it.append(k)
            e.append(error)
            xo = xi
            xi = xn

        k += 1
    matplotlib.pyplot.plot(it, e)
    ans = {
        "x": xn,
        "k": k,
        "error": e[-1]
    }
    return ans

# secante('exp(-(x**2))-x',0,1,10**-8,3)

File: octave/ecuaciones_no_lineales/codigos/secante.m
function [xn, k, error] = secante(f, xo, xi, tol, iterMax)    
    syms x;
    f1 = matlabFunction(sym(f));
    k = 0;
    error = tol + 1 ;
    e = [];
    while (tol < error && k < iterMax)
        num = (xi - xo);
        deno = (f1(xi) - f1(xo));
        if (deno == 0)
            x = [];
            k = [];
            e = [];
            disp('La función se indefine.')
            break;
        else
            xn = xi - ((num./deno)* f1(xi));
            error = abs(xn);
            e = [e error];
            xo = xi;
            xi = xn;
            ++k;
        end
    end
    plot(1:k, e)
    xlabel('iter (k)')
    ylabel('Error (|f(x_k)|)')
    title('Error del Metodo de la Secante')
end

% [x,k,error] = secante(exp(-(x.^2))-x,0,1,10^-8,3)